src/audio/dai-zephyr.c

```diff
@@ -268,16 +268,55 @@ dai_dma_cb(struct dai_data *dd, struct comp_dev *dev, uint32_t bytes,
 	}
 
 	if (dev->direction == SOF_IPC_STREAM_PLAYBACK) {
+#if CONFIG_IPC_MAJOR_4
+		struct list_item *sink_list;
+		/*
+		 * copy from local buffer to all sinks that are not gateway buffers
+		 * using the right PCM converter function.
+		 */
+		list_for_item(sink_list, &dev->bsink_list) {
+			struct comp_dev *sink_dev;
+			struct comp_buffer *sink;
+			int j;
+
+			sink = container_of(sink_list, struct comp_buffer, source_list);
+
+			if (sink == dd->dma_buffer)
+				continue;
+
+			sink_dev = sink->sink;
+
+			j = IPC4_SRC_QUEUE_ID(buf_get_id(sink));
+
+			if (j >= IPC4_COPIER_MODULE_OUTPUT_PINS_COUNT) {
+				comp_err(dev, "Sink queue ID: %d >= max output pin count: %d\n",
+					 j, IPC4_COPIER_MODULE_OUTPUT_PINS_COUNT);
+				ret = -EINVAL;
+				continue;
+			}
+
+			if (!converter[j]) {
+				comp_err(dev, "No PCM converter for sink queue %d\n", j);
+				ret = -EINVAL;
+				continue;
+			}
+
+			if (sink_dev && sink_dev->state == COMP_STATE_ACTIVE) {
```

Comment:
Those additional sinks are (always?) connected to other pipelines. And so some other pipelines might be responsible to set buffer stream params in .params() handler of components connected to those sinks. For unknown reason checking sink_dev->state == COMP_STATE_ACTIVE is not enough, I recently observed problems with multiple-pause-resume test on my draft PR and have to fix it like this: https://github.com/thesofproject/sof/pull/9238/commits/4ebd807a1733f74befe07da64be50beabc6b8cd7.

Frame format of uninitialized buffer is 0 which is treated as SOF_IPC_FRAME_S16_LE. If test uses 24 or 32 bit formats that could lead to wrong consume/produce and invalidate/writeback values.

The problem could be fixed in a separate PR, I just comment it here as your PR adds new place there the fix should be applied.

Reply:
Sure, thanks for review. I'll take a look at your comments and get back with answers